sklearn/preprocessing/tests/test_target_encoder.py

```diff
@@ -0,0 +1,252 @@
+import pytest
+import numpy as np
+from numpy.testing import assert_allclose
+
+from sklearn.preprocessing import TargetRegressorEncoder
+
+
+@pytest.mark.parametrize('categories, unknown_X', [
+    (
+        np.array([0, 1, 2], dtype=int),
+        np.array([4], dtype=int)
+    ),
+    (
+        np.array(['cat', 'dog', 'snake'], dtype=object),
+        np.array(['cow'], dtype=object)
+    ),
+])
+@pytest.mark.parametrize('seed', range(3))
+def test_regression(categories, unknown_X, seed):
+    # checks impact encoder for regression
+
+    X_int = np.array([[0] * 20 + [1] * 30 + [2] * 40], dtype=int).T
+    X_input = categories[X_int]
+    n_samples = X_input.shape[0]
+    rng = np.random.RandomState(seed)
+    y = rng.uniform(low=-10, high=20, size=n_samples)
+
+    # multilevel partial pooling
+    y_sections = [y[:20], y[20:50], y[50:]]
+    cat_means = np.array([np.mean(sect) for sect in y_sections])
+    cat_variance = np.array([np.var(sect) for sect in y_sections])
+    cat_counts = np.array([20, 30, 40])
+
+    y_variance = np.var(y)
+    y_mean = np.mean(y)
+
+    # multilevel partial pooling directly
+    cat_encoded = ((cat_counts * cat_means / cat_variance +
+                    y_mean / y_variance))
+    cat_encoded /= (cat_counts / cat_variance + 1 / y_variance)
+
+    # shuffle
+    shuffled_idx = rng.permutation(n_samples)
+    X_input = X_input[shuffled_idx]
+    y = y[shuffled_idx]
+
+    enc = TargetRegressorEncoder().fit(X_input, y)
+
+    assert len(enc.cat_encodings_) == 1
+    assert enc.y_mean_ == pytest.approx(y_mean)
+    assert_allclose(enc.cat_encodings_[0], cat_encoded)
+
+    expected_encoding = np.take(cat_encoded, X_int[shuffled_idx, :])
+    X_trans = enc.transform(X_input)
+    assert_allclose(expected_encoding, X_trans)
+
+    # check known test data
+    X_test = np.array([[2, 0, 1]], dtype=int).T
+    X_input = categories[X_test]
+    X_trans = enc.transform(X_input)
+    expected_encoding = cat_encoded[X_test]
+    assert_allclose(expected_encoding, X_trans)
+
+    # unknown
+    X_trans = enc.transform([unknown_X])
+    assert_allclose(X_trans, [[y_mean]])
+
+
+@pytest.mark.parametrize('categories, unknown_X', [
+    (
+        np.array([0, 1, 2], dtype=int),
+        np.array([4], dtype=int)
+    ),
+    (
+        np.array(['cat', 'dog', 'snake'], dtype=object),
+        np.array(['cow'], dtype=object)
+    ),
+])
+def test_zero_variance_group(categories, unknown_X):
+    # The first group is constant results in using the mean for the group
+    # as the encoding
```

Comment:
Sorry I don't understand this sentence

May I suggest

When the target is constant for a given category, the category encoding should correspond to that constant target value.